lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     }
                     cb();
                 });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
```

Comment:
I don't think it's a huge issue and not sure we can do much better, but we have to keep in mind that in case the bucket is non-versioned, we may end up deleting a newer "version" of the object that overwrote an older "version", which ends with neither the old nor the new state (no object at all, where there were always an object before). However, the application is supposed to want to overwrite the old object in that case so shouldn't care about it, and the application aborts the new object which means it doesn't care about it either. But it could become an issue if e.g. the application needs to read the old object in order to write the new object on top, since with the deletion it will not be able to anymore.

I am thinking out loud here, it may be okay for the time being, but maybe longer term we need to have something more clever to get rid of orphaned MPU keys that doesn't delete parts that are already part of a valid object.

Reply:
We discussed this within the team and the problem wasn't clear to us, I think I got it, so i'm trying to rephrase- it, please let me know if I'm wrong:

For example, given a non-versioned bucket.

- We create an MPU for `objectA`, complete it: `objectA` exists with upload id `1`.
- Then we do another MPU for the same object (upload id `2`).
- Then the CompleteMPU fails after cloudserver replaced `objectA` with upload id `2`.
- Before the abort, we replace the object again (with a PUT or whatever).
- And then we call the abort, and we end up deleting the wrong object.

But in this very last case, that's what we are actually trying to avoid: if there is no object md, or if the upload id does not match, then we do nothing, so nothing is deleted.
Am I missing something?

Reply:
I think @jonathan-gramain is mentioning the case where the user would want to "update the data" (think download/edit/upload) : the first upload fails, but the Abort causes the data to be removed (neither initial not new version are available anymore).... so the application cannot do the update anymore...

Reply:
Created a ticket to keep track of that issue but won't address it in this PR
https://scality.atlassian.net/browse/CLDSRV-579